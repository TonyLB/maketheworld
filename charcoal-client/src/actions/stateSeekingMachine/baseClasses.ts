interface ISSMWrappedAction<D extends Record<string, any>> {
    (dispatch: any, getState: any): Promise<Partial<D>>;
}

export interface ISSMAction<D extends Record<string, any>> {
    (data: D): ISSMWrappedAction<D>
}

export interface ISSMChoiceState<K> {
    key: K;
    stateType: 'CHOICE';
    choices: Array<K>;
}

export interface ISSMAttemptState<K, D extends Record<string, any>> {
    key: K;
    stateType: 'ATTEMPT';
    action: ISSMAction<D>;
    resolve: K;
    reject: K;
}

export interface ISSMHoldCondition<D extends Record<string, any>> {
    (data: D, getState: any): Boolean;
}

export interface ISSMHoldState<K, D extends Record<string, any>> {
    key: K;
    stateType: 'HOLD';
    condition: ISSMHoldCondition<D>;
    next: K;
}

export type ISSMPotentialState<K, D extends Record<string, any>> = ISSMChoiceState<K> | ISSMAttemptState<K, D> | ISSMHoldState<K, D>

export interface ISSMTemplateAbstract<K extends string, D extends Record<string, any>> {
    initialState: K;
    initialData: D;
    states: Record<K, ISSMPotentialState<K, D>>;
}

export interface IStateSeekingMachineAbstract<K extends string, D extends Record<string, any>, T extends ISSMTemplateAbstract<K, D>> {
    key: string;
    template: T;
    currentState: K;
    desiredState: K;
    data: D;
}
