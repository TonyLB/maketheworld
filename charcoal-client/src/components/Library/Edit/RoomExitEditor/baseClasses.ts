export type RoomExit = {
    key: string;
    to: string;
    from: string;
    name: string;
}
