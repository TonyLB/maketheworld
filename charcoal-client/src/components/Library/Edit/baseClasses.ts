import { BaseEditor, Selection } from 'slate'
import { ReactEditor } from 'slate-react'

export type CustomText = {
    text: string;
}

export type CustomLineElement = {
    type: 'line';
    children: CustomText[]
}

export type CustomActionLinkElement = {
    type: 'actionLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomFeatureLinkElement = {
    type: 'featureLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomLinkElement = CustomActionLinkElement | CustomFeatureLinkElement

export type CustomLineBreak = {
    type: 'lineBreak';
}

export type CustomParagraphContents = CustomText | CustomActionLinkElement | CustomFeatureLinkElement | CustomLineBreak

export const isCustomLineBreak = (item: CustomParagraphContents): item is CustomLineBreak => ('type' in item && item.type === 'lineBreak')
export const isCustomActionLink = (item: CustomParagraphContents): item is CustomActionLinkElement => ('type' in item && item.type === 'actionLink')
export const isCustomFeatureLink = (item: CustomParagraphContents): item is CustomFeatureLinkElement => ('type' in item && item.type === 'featureLink')
export const isCustomLink = (item: CustomParagraphContents): item is CustomLinkElement => (isCustomActionLink(item) || isCustomFeatureLink(item))
export const isCustomText = (item: CustomParagraphContents): item is CustomText => ('text' in item)

export type CustomParagraphElement = {
    type: 'paragraph';
    children: CustomParagraphContents[]
}

export type CustomDescriptionElement = {
    type: 'description';
    children: CustomParagraphElement[]
}

type CustomElement = CustomLineElement |
    CustomActionLinkElement |
    CustomFeatureLinkElement |
    CustomParagraphElement |
    CustomDescriptionElement



declare module 'slate' {
    interface CustomTypes {
      Editor: BaseEditor & ReactEditor & { saveSelection?: Selection }
      Element: CustomElement
      Text: CustomText
    }
  }

export {}