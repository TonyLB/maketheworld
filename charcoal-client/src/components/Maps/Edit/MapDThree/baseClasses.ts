import {
    SimulationNodeDatum,
    SimulationLinkDatum
} from 'd3-force'

export type SimNode = SimulationNodeDatum & {
    id: string;
    zLevel?: number;
    cascadeNode: boolean;
    roomId: string;
    visible: boolean;
}
export type NodeRecord = Record<string, SimNode>
export type LinkRecord = {
    id: string,
    source: string,
    target: string,
    visible?: boolean
}[]
export type SimulationReturn = {
    key: string,
    nodes: SimNode[],
    links: SimulationLinkDatum<SimNode>[]
}

export type SimCallback = (nodes: SimNode[]) => void

export type MapNodes = SimNode[]
export type MapLinks = SimulationLinkDatum<SimulationNodeDatum>[]
