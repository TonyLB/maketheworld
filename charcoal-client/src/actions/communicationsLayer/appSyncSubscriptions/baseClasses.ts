import { ISSMTemplateAbstract, ISSMPotentialState, ISSMAction, ISSMHoldCondition } from '../../stateSeekingMachine/baseClasses'

export const SUBSCRIPTION_SUCCESS = 'SUBSCRIPTION_SUCCESS'

export type subscriptionSSMKeys = 'INITIAL' | 'AVAILABLE' | 'SUBSCRIBING' | 'SUBSCRIBED' | 'SYNCHING' | 'SYNCHRONIZED' | 'UNSUBSCRIBING'
export type cachedSubscriptionSSMKeys = 'INITIAL' | 'FETCHING' | 'FETCHED' | 'SUBSCRIBING' | 'SUBSCRIBED' | 'SYNCHING' | 'SYNCHRONIZED' | 'UNSUBSCRIBING'

//
// Implement a state-seeking machine to keep websockets connected where possible.
//
export const subscriptionSSMClassGenerator = <D extends Record<string, any>, T extends string>({
    ssmType, initialData, condition, subscribeAction, unsubscribeAction, syncAction
}: {
    ssmType: T;
    initialData: D;
    condition: ISSMHoldCondition<D>;
    subscribeAction: ISSMAction<D>;
    unsubscribeAction: ISSMAction<D>;
    syncAction: ISSMAction<D>
}) => class implements ISSMTemplateAbstract<subscriptionSSMKeys, D> {
    ssmType: T = ssmType
    initialState: subscriptionSSMKeys = 'INITIAL'
    initialData: D = initialData
    states: Record<subscriptionSSMKeys, ISSMPotentialState<subscriptionSSMKeys, D>> = {
        INITIAL: {
            stateType: 'CHOICE',
            key: 'INITIAL',
            choices: ['AVAILABLE']
        },
        AVAILABLE: {
            stateType: 'HOLD',
            key: 'AVAILABLE',
            condition,
            next: 'SUBSCRIBING'
        },
        SUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'SUBSCRIBING',
            action: subscribeAction,
            resolve: 'SUBSCRIBED',
            reject: 'INITIAL'
        },
        SUBSCRIBED: {
            stateType: 'CHOICE',
            key: 'SUBSCRIBED',
            choices: ['UNSUBSCRIBING', 'SYNCHING']
        },
        UNSUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'UNSUBSCRIBING',
            action: unsubscribeAction,
            resolve: 'INITIAL',
            reject: 'INITIAL'
        },
        SYNCHING: {
            stateType: 'ATTEMPT',
            key: 'SYNCHING',
            action: syncAction,
            resolve: 'SYNCHRONIZED',
            reject: 'SUBSCRIBED'
        },
        SYNCHRONIZED: {
            stateType: 'CHOICE',
            key: 'SYNCHRONIZED',
            choices: ['UNSUBSCRIBING']
        },
    }
}

//
// Implement a more elaborate state-seeking machine to keep websockets connected where possible,
// including fetching cached information.
//
export const cachedSubscriptionSSMClassGenerator = <D extends Record<string, any>, T extends string>({
    ssmType, initialData, fetchAction, subscribeAction, unsubscribeAction, syncAction
}: {
    ssmType: T;
    initialData: D;
    fetchAction: ISSMAction<D>;
    subscribeAction: ISSMAction<D>;
    unsubscribeAction: ISSMAction<D>;
    syncAction: ISSMAction<D>
}) => class implements ISSMTemplateAbstract<cachedSubscriptionSSMKeys, D> {
    ssmType: T = ssmType
    initialState: cachedSubscriptionSSMKeys = 'INITIAL'
    initialData: D = initialData
    states: Record<cachedSubscriptionSSMKeys, ISSMPotentialState<cachedSubscriptionSSMKeys, D>> = {
        INITIAL: {
            stateType: 'CHOICE',
            key: 'INITIAL',
            choices: ['FETCHING']
        },
        FETCHING: {
            stateType: 'ATTEMPT',
            key: 'FETCHING',
            action: fetchAction,
            resolve: 'FETCHED',
            reject: 'INITIAL'
        },
        FETCHED: {
            stateType: 'CHOICE',
            key: 'FETCHED',
            choices: ['SUBSCRIBING']
        },
        SUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'SUBSCRIBING',
            action: subscribeAction,
            resolve: 'SUBSCRIBED',
            reject: 'INITIAL'
        },
        SUBSCRIBED: {
            stateType: 'CHOICE',
            key: 'SUBSCRIBED',
            choices: ['UNSUBSCRIBING', 'SYNCHING']
        },
        UNSUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'UNSUBSCRIBING',
            action: unsubscribeAction,
            resolve: 'INITIAL',
            reject: 'INITIAL'
        },
        SYNCHING: {
            stateType: 'ATTEMPT',
            key: 'SYNCHING',
            action: syncAction,
            resolve: 'SYNCHRONIZED',
            reject: 'SUBSCRIBED'
        },
        SYNCHRONIZED: {
            stateType: 'CHOICE',
            key: 'SYNCHRONIZED',
            choices: ['UNSUBSCRIBING']
        }
    }
}
