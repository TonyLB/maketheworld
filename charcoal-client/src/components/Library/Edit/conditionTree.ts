//
// TODO: Create converter function that reduces a list of exits into a tree of nodes with (a) a list of
// exits at that node, and (b) a list of If clusters (1 If, zero or more ElseIfs, zero or one Else), each
// their own node.
//

import { isNormalCondition, isNormalExit, NormalCondition, NormalConditionStatement, NormalForm, NormalItem, NormalReference } from "@tonylb/mtw-wml/dist/normalize/baseClasses";
import { deepEqual } from "../../../lib/objects";
import { addItem } from "../../../slices/activeCharacters";
import { noConditionContext } from "./utilities";

type ConditionalTreeNode<T extends NormalItem> = {
    if: {
        source: string;
        node: ConditionalTree<T>;
    };
    elseIfs: {
        source: string;
        node: ConditionalTree<T>;
    }[];
    else?: ConditionalTree<T>;
}

export type ConditionalTree<T extends NormalItem> = {
    items: T[];
    conditionals: ConditionalTreeNode<T>[];
}

const mergeItem = <T extends NormalItem>(compare: (A: T, B: T) => boolean ) => (previous: (T)[], item: T): (T)[] => {
    if (previous.find((check) => (compare(check, item)))) {
        return previous
    }
    else {
        return [
            ...previous,
            item
        ]
    }
}

type MatchExistingTreeOutput = {
    type: 'If';
} | {
    type: 'ElseIf';
    index: number;
} | {
    type: 'NewElseIf';
    additionalConditions: string[];
} | {
    type: 'Else';
} | {
    type: 'NewElse';
    additionalConditions: string[];
}

const matchExistingTree = <T extends NormalItem>(node: ConditionalTreeNode<T>, { conditions }: NormalCondition): MatchExistingTreeOutput | undefined => {
    const hasSource = Boolean(conditions.find(({ not }) => (!not)))
    const extendsConditions = conditions.length > node.elseIfs.length + 1
    if (extendsConditions) {
        //
        // The existing chain of elseIfs must all match against the incoming conditions, or there is no match
        //
        if (!deepEqual(conditions.slice(0, node.elseIfs.length + 1).map(({ if: source }) => (source)), [node.if.source, ...node.elseIfs.map(({ source }) => (source))])) {
            return undefined
        }
        const additionalConditions = conditions.slice(node.elseIfs.length + 1).map(({ if: source }) => (source))
        if (hasSource) {
            return { type: 'NewElseIf', additionalConditions }
        }
        else {
            return { type: 'NewElse', additionalConditions }
        }
    }
    else {
        //
        // The incoming conditions must all match against the existing chain of elseIfs, or there is no match
        //
        if (!deepEqual(conditions.map(({ if: source }) => (source)), [node.if.source, ...node.elseIfs.map(({ source }) => (source))].slice(0, conditions.length))) {
            return undefined
        }
        if (conditions.length === 1 && hasSource) {
            return { type: 'If' }
        }
        if (!hasSource && conditions.length < node.elseIfs.length) {
            //
            // Cannot break off for an unconditioned Else in the middle of this ElseIf chain
            //
            return undefined
        }
        if (hasSource) {
            return { type: 'ElseIf', index: conditions.length - 1 }
        }
        else {
            return { type: 'Else' }
        }
    }
}

const addItemToTreeInContext = <T extends NormalItem>(compare: (A: T, B: T) => boolean, normalForm: NormalForm ) => (previous: ConditionalTree<T>, { item, contextStack }: { item: T, contextStack: NormalReference[] }): ConditionalTree<T> => {
    if (contextStack.length)  {
        const firstCondition = normalForm[contextStack[0].key]
        if (firstCondition && isNormalCondition(firstCondition)) {
            const matchRecord = previous.conditionals.reduce<{ index: number; match: MatchExistingTreeOutput } | undefined>((accumulator, node, index) => {
                if (!accumulator) {
                    const match = matchExistingTree(node, firstCondition)
                    if (match) {
                        return { index, match }
                    }
                }
                return accumulator
            }, undefined)
            const newArg = { item, contextStack: contextStack.slice(1) }
            if (matchRecord) {
                //
                // Add new conditions to the current node if needed
                //
                const { index, match } = matchRecord
                let newNode = previous.conditionals[index]
                switch(match.type) {
                    case 'If':
                        newNode = {
                            ...newNode,
                            if: {
                                ...newNode.if,
                                node: addItemToTreeInContext(compare, normalForm)(newNode.if.node, newArg)
                            }
                        }
                        break
                    case 'Else':
                        newNode = {
                            ...newNode,
                            else: addItemToTreeInContext(compare, normalForm)(newNode.else ?? { items: [], conditionals: [] }, newArg)
                        }
                        break
                    case 'ElseIf':
                        newNode = {
                            ...newNode,
                            elseIfs: [
                                ...newNode.elseIfs.slice(0, match.index),
                                {
                                    ...newNode.elseIfs[match.index],
                                    node: addItemToTreeInContext(compare, normalForm)(newNode.elseIfs[match.index].node, newArg)
                                }
                            ]
                        }
                        break
                    case 'NewElse':
                        newNode = {
                            ...newNode,
                            elseIfs: [
                                ...newNode.elseIfs,
                                ...match.additionalConditions.map((source) => ({ source, node: { items: [], conditionals: []  }}))
                            ],
                            else: addItemToTreeInContext(compare, normalForm)({ items: [], conditionals: [] }, newArg)
                        }
                        break
                    case 'NewElseIf':
                        newNode = {
                            ...newNode,
                            elseIfs: [
                                ...newNode.elseIfs,
                                ...match.additionalConditions.slice(0, -1).map((source) => ({ source, node: { items: [], conditionals: []  }})),
                                {
                                    source: match.additionalConditions.slice(-1)[0],
                                    node: addItemToTreeInContext(compare, normalForm)({ items: [], conditionals: [] }, newArg)
                                }
                            ]
                        }
                }
                return {
                    ...previous,
                    conditionals: [
                        ...previous.conditionals.slice(0, index),
                        newNode,
                        ...previous.conditionals.slice(index + 1)
                    ]
                }
            }
            else {
                //
                // Create a new condition node with matching structures
                //
                const { conditions } = firstCondition
                if (conditions.length) {
                    const hasSource = Boolean(conditions.find(({ not }) => (!not)))
                    let newNode: ConditionalTreeNode<T>
                    if (hasSource) {
                        if (conditions.length > 1) {
                            newNode = {
                                if: {
                                    source: conditions[0].if,
                                    node: {
                                        items: [],
                                        conditionals: []
                                    }
                                },
                                elseIfs: [
                                    ...conditions.slice(1, -1).map(({ if: ifPredicate }) => ({
                                        source: ifPredicate,
                                        node: {
                                            items: [],
                                            conditionals: []
                                        }
                                    })),
                                    {
                                        source: conditions.slice(-1)[0].if,
                                        node: addItemToTreeInContext(compare, normalForm)({ items: [], conditionals: [] }, newArg)
                                    }
                                ]
                            }
                        }
                        else {
                            newNode = {
                                if: {
                                    source: conditions[0].if,
                                    node: addItemToTreeInContext(compare, normalForm)({ items: [], conditionals: [] }, newArg)
                                },
                                elseIfs: []
                            }
                        }
                    }
                    else {
                        newNode = {
                            if: {
                                source: conditions[0].if,
                                node: {
                                    items: [],
                                    conditionals: []
                                }
                            },
                            elseIfs: conditions.slice(1).map(({ if: ifPredicate }) => ({
                                source: ifPredicate,
                                node: {
                                    items: [],
                                    conditionals: []
                                }
                            })),
                            else: addItemToTreeInContext(compare, normalForm)({ items: [], conditionals: [] }, newArg)
                        }
                    }
                    return {
                        ...previous,
                        conditionals: [
                            ...previous.conditionals,
                            newNode
                        ]
                    }
                }
            }
        }
        return previous
    }
    else {
        return {
            ...previous,
            items: mergeItem(compare)(previous.items, item)
        }
    }
}

const recursiveSimplifyConditionTree = <T extends NormalItem>(tree: ConditionalTree<T>): any => ({
    items: tree.items.map((item) => (isNormalExit(item) ? item.name : '')),
    conditionals: tree.conditionals.map(({ if: ifPredicate, elseIfs, else: elsePredicate }) => ({
        if: {
            source: ifPredicate.source,
            node: recursiveSimplifyConditionTree(ifPredicate.node)
        },
        elseIfs: elseIfs.map(({ source, node }) => ({
            source,
            node: recursiveSimplifyConditionTree(node)
        })),
        else: elsePredicate ? recursiveSimplifyConditionTree(elsePredicate) : undefined
    }))
})

export const reduceItemsToTree = <T extends NormalItem>(compare: (A: T, B: T) => boolean, normalForm: NormalForm) => (previous: ConditionalTree<T>, item: T): ConditionalTree<T> => {
    if ((item.appearances ?? []).find(noConditionContext)) {
        return {
            ...previous,
            items: mergeItem(compare)(previous.items, item)
        }
    }
    else {
        return (item.appearances ?? [])
            .map(({ contextStack }) => ({ item, contextStack: contextStack.filter(({ tag }) => (tag === 'If')) }))
            .reduce<ConditionalTree<T>>((previous, item) => (addItemToTreeInContext(compare, normalForm)(previous, item)), previous)
    }
}
