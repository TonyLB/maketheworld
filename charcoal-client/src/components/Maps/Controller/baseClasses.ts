import { GenericTree } from "@tonylb/mtw-sequence/dist/tree/baseClasses"
import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition

type MapContextExitDrag = {
    sourceRoomId: string;
    x: number;
    y: number;
}

type MapDispatchSetTool = {
    type: 'SetToolSelected';
    value: ToolSelected;
}

type MapDispatchSetExitDrag = {
    type: 'SetExitDrag';
} & Partial<MapContextExitDrag>

export type MapDispatchAction = MapDispatchSetTool | MapDispatchSetExitDrag

export type MapContextType = {
    mapId: string;
    tree: GenericTree<MapTreeItem>;
    UI: {
        //
        // The Map editor can conceivably need data for:
        //    - Which tool is selected in the toolbar
        //    - Whether an exit is being dragged, from where, and to where
        // Updates to this data should be performed through the mapDispatch
        // function.
        //
        toolSelected: ToolSelected;
        exitDrag: MapContextExitDrag;
    },
    // mapD3: MapDThree,
    mapDispatch: (action: MapDispatchAction) => void;
}
