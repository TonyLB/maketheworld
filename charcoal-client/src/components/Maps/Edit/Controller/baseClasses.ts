import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean; contents: MapTreeItem[] }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean; contents: MapTreeItem[] }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition
