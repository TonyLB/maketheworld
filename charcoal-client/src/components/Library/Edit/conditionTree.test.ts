import Normalizer from "@tonylb/mtw-wml/dist/normalize"
import { isNormalExit, NormalExit, NormalForm } from "@tonylb/mtw-wml/dist/normalize/baseClasses"
import parse from "@tonylb/mtw-wml/dist/parser"
import tokenizer from "@tonylb/mtw-wml/dist/parser/tokenizer"
import SourceStream from "@tonylb/mtw-wml/dist/parser/tokenizer/sourceStream"
import { schemaFromParse } from "@tonylb/mtw-wml/dist/schema"
import { ConditionalTree, reduceItemsToTree } from "./conditionTree"

const recursiveSimplifyConditionTree = (tree: ConditionalTree<NormalExit>): any => ({
    items: tree.items.map(({ name }) => (name)),
    conditionals: tree.conditionals.map(({ if: ifPredicate, elseIfs, else: elsePredicate }) => ({
        if: {
            source: ifPredicate.source,
            node: recursiveSimplifyConditionTree(ifPredicate.node)
        },
        elseIfs: elseIfs.map(({ source, node }) => ({
            source,
            node: recursiveSimplifyConditionTree(node)
        })),
        else: elsePredicate ? recursiveSimplifyConditionTree(elsePredicate) : undefined
    }))
})

describe('conditionTree', () => {
    const reducer = (normalForm: NormalForm) => (reduceItemsToTree({ compare: (A: { key: string; name: string; }, B: { key: string; name: string; }): boolean => (A.key === B.key), normalForm, transform: ({ key, name }: NormalExit) => ({ key, name: name || '' }) }))
    const normalFromSource = (source: string) => {
        const normalizer = new Normalizer()
        const testCharacter = schemaFromParse(parse(tokenizer(new SourceStream(source))))
        normalizer.put(testCharacter[0], { contextStack: [], index: 0, replace: false })
        return normalizer.normal
    }

    it('should create nested elseIf items', () => {
        const testNormalForm = normalFromSource(`<Asset key=(test)>
            <Room key=(targetOne) />
            <Room key=(targetTwo) />
            <Room key=(targetThree) />
            <Room key=(targetFour) />
            <Variable key=(varOne) default={false} />
            <Variable key=(varTwo) default={false} />
            <Variable key=(varThree) default={false} />
            <Room key=(testRoom)>
                <If {varOne}>
                    <If {varTwo}>
                        <Exit to=(targetTwo)>target two</Exit>
                    </If>
                    <ElseIf {varThree}>
                        <Exit to=(targetThree)>target three</Exit>
                    </ElseIf>
                </If>
            </Room>
        </Asset>`)
        const testTree = Object.values(testNormalForm).filter(isNormalExit).reduce<ConditionalTree<{ key: string; name: string; }>>(reducer(testNormalForm), { items: [], conditionals: [] })
        expect(testTree).toMatchSnapshot()
    })

    it('should create complete tree given all exits', () => {
        const testNormalForm = normalFromSource(`<Asset key=(test)>
            <Room key=(targetOne) />
            <Room key=(targetTwo) />
            <Room key=(targetThree) />
            <Room key=(targetFour) />
            <Variable key=(varOne) default={false} />
            <Variable key=(varTwo) default={false} />
            <Variable key=(varThree) default={false} />
            <Room key=(testRoom)>
                <Exit to=(targetOne)>target one</Exit>
                <If {varOne}>
                    <If {varTwo}>
                        <Exit to=(targetTwo)>target two</Exit>
                    </If>
                    <ElseIf {varThree}>
                        <Exit to=(targetThree)>target three</Exit>
                    </ElseIf>
                    <Else>
                        <Exit to=(targetFour)>target four</Exit>
                    </Else>
                </If>
            </Room>
            <If {varOne}>
                <Room key=(targetOne)><Exit to=(testRoom)>fromOne</Exit></Room>
            </If>
            <Else>
                <Room key=(targetTwo)><Exit to=(testRoom)>fromTwo</Exit></Room>
            </Else>
        </Asset>`)
        const testTree = Object.values(testNormalForm).filter(isNormalExit).reduce<ConditionalTree<{ key: string; name: string; }>>(reducer(testNormalForm), { items: [], conditionals: [] })
        expect(testTree).toMatchSnapshot()
    })
})