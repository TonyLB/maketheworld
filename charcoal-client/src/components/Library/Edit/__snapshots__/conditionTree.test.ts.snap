// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditionTree should create complete tree given all exits 1`] = `
Object {
  "conditionals": Array [
    Object {
      "else": Object {
        "conditionals": Array [],
        "items": Array [
          "fromTwo",
        ],
      },
      "elseIfs": Array [],
      "if": Object {
        "node": Object {
          "conditionals": Array [
            Object {
              "else": Object {
                "conditionals": Array [],
                "items": Array [
                  "target four",
                ],
              },
              "elseIfs": Array [
                Object {
                  "node": Object {
                    "conditionals": Array [],
                    "items": Array [
                      "target three",
                    ],
                  },
                  "source": "varThree",
                },
              ],
              "if": Object {
                "node": Object {
                  "conditionals": Array [],
                  "items": Array [
                    "target two",
                  ],
                },
                "source": "varTwo",
              },
            },
          ],
          "items": Array [
            "fromOne",
          ],
        },
        "source": "varOne",
      },
    },
  ],
  "items": Array [
    "target one",
  ],
}
`;

exports[`conditionTree should create nested elseIf items 1`] = `
Object {
  "conditionals": Array [
    Object {
      "else": undefined,
      "elseIfs": Array [],
      "if": Object {
        "node": Object {
          "conditionals": Array [
            Object {
              "else": undefined,
              "elseIfs": Array [
                Object {
                  "node": Object {
                    "conditionals": Array [],
                    "items": Array [
                      "target three",
                    ],
                  },
                  "source": "varThree",
                },
              ],
              "if": Object {
                "node": Object {
                  "conditionals": Array [],
                  "items": Array [
                    "target two",
                  ],
                },
                "source": "varTwo",
              },
            },
          ],
          "items": Array [],
        },
        "source": "varOne",
      },
    },
  ],
  "items": Array [],
}
`;
